Mathlib/Topology/ContinuousFunction/CompactlySupported.lean

```diff
@@ -0,0 +1,551 @@
+    SMulCommClass R C_c(α, β) C_c(α, β) where
+  smul_comm r f g := by
+    ext
+    simp only [smul_eq_mul, coe_smul, coe_mul, Pi.smul_apply, Pi.mul_apply]
+    rw [← smul_eq_mul, ← smul_eq_mul, smul_comm]
+
+end AlgebraicStructure
+
+section Uniform
+
+variable [UniformSpace β] [UniformSpace γ] [Zero γ]
+variable [FunLike F β γ] [CompactlySupportedContinuousMapClass F β γ]
+
+theorem uniformContinuous (f : F) : UniformContinuous (f : β → γ) :=
+  (map_continuous f).uniformContinuous_of_tendsto_cocompact
+  (HasCompactSupport.is_zero_at_infty (hasCompactSupport f))
+
+end Uniform
+
+section Star
+
+/-! ### Star structure
+
+It is possible to equip `C_c(α, β)` with a pointwise `star` operation whenever there is a continuous
+`star : β → β` for which `star (0 : β) = 0`. We don't have quite this weak a typeclass, but
+`StarAddMonoid` is close enough.
+
+The `StarAddMonoid` and `NormedStarGroup` classes on `C_c(α, β)` are inherited from their
+counterparts on `α →ᵇ β`. Ultimately, when `β` is a C⋆-ring, then so is `C_c(α, β)`.
+-/
+
+
+variable [TopologicalSpace β] [AddMonoid β] [StarAddMonoid β] [ContinuousStar β]
+
+instance : Star C_c(α, β) where
+  star f :=
+    { toFun := fun x => star (f x)
+      continuous_toFun := (map_continuous f).star
+      hasCompactSupport' := by
+        rw [HasCompactSupport, tsupport]
+        simp only
+        have support_star : (Function.support fun (x : α) => star (f x)) = Function.support f := by
+          ext x
+          simp only [Function.mem_support, ne_eq, star_eq_zero]
+        rw [support_star]
+        exact f.2 }
+
+@[simp]
+theorem coe_star (f : C_c(α, β)) : ⇑(star f) = star (⇑f) :=
+  rfl
+
+theorem star_apply (f : C_c(α, β)) (x : α) : (star f) x = star (f x) :=
+  rfl
+
+instance instStarAddMonoid [ContinuousAdd β] : StarAddMonoid C_c(α, β) where
+  star_involutive f := ext fun x => star_star (f x)
+  star_add f g := ext fun x => star_add (f x) (g x)
+
+end Star
+
+section StarModule
+
+variable {𝕜 : Type*} [Zero 𝕜] [Star 𝕜] [AddMonoid β] [StarAddMonoid β] [TopologicalSpace β]
+  [ContinuousStar β] [SMulWithZero 𝕜 β] [ContinuousConstSMul 𝕜 β] [StarModule 𝕜 β]
+
+instance : StarModule 𝕜 C_c(α, β) where
+  star_smul k f := ext fun x => star_smul k (f x)
+
+end StarModule
+
+section StarRing
+
+variable [NonUnitalSemiring β] [StarRing β] [TopologicalSpace β] [ContinuousStar β]
+  [TopologicalSemiring β]
+
+instance : StarRing C_c(α, β) :=
+  { CompactlySupportedContinuousMap.instStarAddMonoid with
+    star_mul := fun f g => ext fun x => star_mul (f x) (g x) }
+
+end StarRing
+
+/-! ### C_c as a functor
+
+For each `β` with sufficient structure, there is a contravariant functor `C_c(-, β)` from the
+category of topological spaces with morphisms given by `CocompactMap`s.
+-/
+
+
+variable {δ : Type*} [TopologicalSpace β] [TopologicalSpace γ] [TopologicalSpace δ]
+
+local notation α " →co " β => CocompactMap α β
+
+section
+
+variable [T2Space γ] [Zero δ]
```

Comment:
I tried, there is also an option to assume `[R1Space β]`. I don't know whether this works without any assumption. Which one is better? Or both?

```lean
variable [R1Space β] [Zero δ]

/-- Composition of a continuous function with compact support with a cocompact map from an `[R1Space]`
yields another continuous function with compact support. -/
def comp (f : C_c(γ, δ)) (g : β →co γ) : C_c(β, δ) where
  toContinuousMap := (f : C(γ, δ)).comp g
  hasCompactSupport' := by
    simp only [ContinuousMap.toFun_eq_coe, ContinuousMap.coe_comp, ContinuousMap.coe_coe]
    rw [HasCompactSupport]
    have hsuppfc : (Function.support f)ᶜ ∈ Filter.cocompact γ := by
      rw [Filter.mem_cocompact]
      use tsupport f
      constructor
      · exact f.2
      · simp only [Set.compl_subset_compl, Function.support_subset_iff, ne_eq]
        intro x hx
        push_neg at hx
        exact Set.mem_of_subset_of_mem subset_closure (Function.mem_support.mpr hx)
    apply IsCompact.of_isClosed_subset _ isClosed_closure
    rw [Function.support_comp_eq_preimage]
    obtain ⟨K, hK⟩ := Filter.mem_cocompact.mp (g.cocompact_tendsto' hsuppfc)
    simp only [ContinuousMap.toFun_eq_coe, CocompactMap.coe_toContinuousMap, Set.preimage_compl,
      Set.compl_subset_compl] at hK
    exact IsCompact.of_isClosed_subset (IsCompact.closure hK.1) isClosed_closure (closure_mono hK.2)
```

Reply:
Does it work for [R1Space gamma]?

Reply:
I am not sure. I don't think [IsCompact.isClosed](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Separation.html#IsCompact.isClosed) is valid for `[R1Space X]`, I think a counterexample is any compact R1Space `X`, choosing a point `x : X`, artificially adding a point `y` , considering the space `X ∪ {y}`and declaring that `x` and `y` are indistinguishable. `X` is still compact (any open cover has a finite subcover), but `X` is clearly not closed. So I don't know how to use `[R1Space γ]`. Any idea?

Reply:
In fact there is no need for any separation assumption, since the sets we consider (`tsupport _`) are both compact and closed, so their preimages are both compact and closed also, and we never need that the closure of a compact set is compact. I have pushed a proof.

Reply:
very nice, thank you!